### PR TITLE
Rails 6.1: Fix 'TypeError: can't quote ActiveRecord::Relation::QueryAttribute'

### PR DESCRIPTION
**Pull request by mgrunberg**
There are some test failing due to `TypeError: can't quote ActiveRecord::Relation::QueryAttribute`.

I found two places where this was happening:

1. `visit_Arel_Nodes_HomogeneousIn` is binding `ActiveRecord::Relation::QueryAttribute` even for unprepared statements (.i.e `Post.where(author: [1, 2]).to_sql`). Unprepared statement uses [`Arel::Collectors::SubstituteBinds`](https://github.com/rails/rails/blob/v6.1.0/activerecord/lib/arel/collectors/substitute_binds.rb) which quote the binded value (see [abstract_adapter](https://github.com/rails/rails/blob/v6.1.0/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L729) to confirm what I'm saying)
2. explain receives instances of `::ActiveModel::Attribute` (not only by `visit_Arel_Nodes_HomogeneousIn` hack).

### CODE REVIEW

File: lib/active_record/connection_adapters/sqlserver/core_ext/explain.rb [hunk 1]
```diff
@@ -27,7 +27,12 @@ def unprepare_sqlserver_statement(sql, binds)
             executesql = executesql.match(SQLSERVER_STATEMENT_REGEXP).to_a[1]
 
             binds.each_with_index do |bind, index|
-              value = connection.quote(bind)
+
+              value = if ::ActiveModel::Attribute === bind then
+                connection.quote(bind.value_for_database)
+              else
+                connection.quote(bind)
+              end
```
Comment (wpolicarpo) [inline lib/active_record/connection_adapters/sqlserver/core_ext/explain.rb:35]:
```ruby
value = if bind.is_a?(::ActiveModel::Attribute)
          connection.quote(bind.value_for_database)
        else
          connection.quote(bind)
        end
```

**Comment by wpolicarpo** [comment]
Can you please add a changelog entry for this one?

**Comment by aidanharan** [comment]
@mgrunberg I have a cleaner fix for this issue in https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pull/885. Could you review and see what you think?

**Comment by mgrunberg** [comment]
> @mgrunberg I have a cleaner fix for this issue in #885. Could you review and see what you think?

@aidanharan I evaluated adding `ActiveModel::Attribute` to _quote but discarded it after a while. Here is what I thought,

Rails deprecated passing ActiveRecord objects to quote (https://github.com/rails/rails/commit/87886c93dc22d528d8529d8b63873f699beaf695). I consider passing ActiveModel::Attributes as something similar to the deprecated behavior. The fact that we call `quote` inside `_quote` is a smell of mixing a lower-level API with a higher one. 

Allowing quoting of `ActiveModel::Attributes` but not typecasting seems inconsistent to me.

Digging in Rails code (this time main branch) I keep finding places where they decide not to quote `ActiveModel::Attributes` directly (https://github.com/rails/rails/blob/main/activerecord/lib/arel/visitors/to_sql.rb#L106, https://github.com/rails/rails/blob/main/activerecord/lib/arel/visitors/to_sql.rb#L612)

What do you think?

**Comment by aidanharan** [comment]
@mgrunberg Yeah I think your fix is better. I've created https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pull/896 to revert https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pull/885 so that this PR can be used instead.

**Comment by wpolicarpo** [comment]
Hey, can you resolve the changelog conflict so I can merge it?

**Comment by mgrunberg** [comment]
@wpolicarpo done

**Pull request merged**